### PR TITLE
Update train.py

### PR DESCRIPTION
**Pull request by ntyoshi**
Fixed the saving process to reflect pytorch-lightning upgrading
#162

### CODE REVIEW

File: egs/wham/DPRNN/train.py [hunk 1]
```diff
@@ -66,30 +66,26 @@ def main(conf):
 
     # Define callbacks
     checkpoint_dir = os.path.join(exp_dir, 'checkpoints/')
-    checkpoint = ModelCheckpoint(checkpoint_dir, monitor='val_loss',
-                                 mode='min', save_top_k=5, verbose=1)
+    checkpoint = ModelCheckpoint(checkpoint_dir, monitor='val_loss', verbose=True,
+                                 save_last=False, save_top_k=5, save_weights_only=False, 
+                                 mode='min', period=1, prefix='')
```
Comment (mpariente) [inline egs/wham/DPRNN/train.py:71]: You don't need to write the default values if you don't change them. 
```suggestion
    checkpoint = ModelCheckpoint(checkpoint_dir, monitor='val_loss', verbose=True,
                                 save_top_k=5')
```

File: egs/wham/DPRNN/train.py [hunk 2]
```diff
@@ -66,30 +66,26 @@ def main(conf):
...
     trainer.fit(system)
 
-    best_k = {k: v.item() for k, v in checkpoint.best_k_models.items()}
-    with open(os.path.join(exp_dir, "best_k_models.json"), "w") as f:
-        json.dump(best_k, f, indent=0)
-
-    # Save best model (next PL version will make this easier)
-    best_path = [b for b, v in best_k.items() if v == min(best_k.values())][0]
+    best_path = checkpoint.best_model_path
     state_dict = torch.load(best_path)
```
Comment (mpariente) [inline egs/wham/DPRNN/train.py:89]: Can you merge those two lines please?

File: egs/wham/DPRNN/train.py (hunk 1 above)
Comment (ntyoshi) [reply egs/wham/DPRNN/train.py:71]: I know. Do you prefer to delete the default values?

File: egs/wham/DPRNN/train.py (hunk 2 above)
Comment (ntyoshi) [reply egs/wham/DPRNN/train.py:89]: OK. I will.

**Comment by ntyoshi** [comment]
@mpariente 
Should I create another PR that will be fixed the things you commented?

File: egs/wham/DPRNN/train.py (hunk 1 above)
Comment (mpariente) [reply egs/wham/DPRNN/train.py:71]: Yes, I think it's clearer in this way.

**Comment by mpariente** [comment]
No, you can push the commits to your fork and the changes will appear here.

**Comment by ntyoshi** [comment]
OK. I will.

File: egs/wham/DPRNN/train.py [hunk 3]
```diff
@@ -66,31 +66,25 @@ def main(conf):
 
     # Define callbacks
     checkpoint_dir = os.path.join(exp_dir, 'checkpoints/')
-    checkpoint = ModelCheckpoint(checkpoint_dir, monitor='val_loss',
-                                 mode='min', save_top_k=5, verbose=1)
+    checkpoint = ModelCheckpoint(checkpoint_dir, monitor='val_loss', verbose=True,
+                                 save_top_k=5')
```
Comment (mpariente) [inline egs/wham/DPRNN/train.py:70]:
```suggestion
    checkpoint = ModelCheckpoint(checkpoint_dir, monitor='val_loss', verbose=True,
                                 mode='min', save_top_k=5)
```

**Comment by mpariente** [comment]
Hey @ntyoshi, any news about this?

**Comment by ntyoshi** [comment]
@mpariente 
In order to check if my fixing is valid, I am trying to train and evaluate the DPRNN model that has around the parameter you mentioned on [README.md](https://github.com/mpariente/asteroid/blob/master/egs/wham/DPRNN/README.md).

When I tried the default values except for kernel_size=2, chunk_size=250 and batch_size=8, I got the result below:
```
 {'sar': -4.556636353964018,
 'sar_imp': -153.73965922673975,
 'sdr': -5.424828011563545,
 'sdr_imp': -5.774611806596595,
 'si_sdr': -6.43818235206604,
 'si_sdr_imp': -6.435335527528233,
 'sir': 10.024114981351794,
 'sir_imp': 9.674331186318728,
 'stoi': 0.6209898268337961,
 'stoi_imp': -0.06155946881815264}
```
I guess `sdr_imp` should be about 19.5 and `si_sdr_imp` should be about 19.3.
Is it true? 

(I am training the same model after I changed batch_size to 3.)

**Comment by mpariente** [comment]
What is your kernel size and chunk size then?

**Comment by ntyoshi** [comment]
I set kernel_size=2 and chunk_size=250 in run.sh.

**Comment by mpariente** [comment]
That sounds very weird indeed.. Have you listened to your data?

**Comment by ntyoshi** [comment]
I've done a few files but let me check more.
Do you have a good idea to check them all?

**Comment by ntyoshi** [comment]
And have you ever heard or experienced such a strange result?

**Comment by mpariente** [comment]
> I've done a few files but let me check more.
> Do you have a good idea to check them all?

No, just listen few of the clean files to be sure that this is indeed clean speech.



> And have you ever heard or experienced such a strange result?

Yes,but not as bad.. 
Can you try with kernel_size 16, this is way more stable than the kernel_size 2. 
Also, did you inspect the log? How is the training loss?

**Comment by ntyoshi** [comment]
>No, just listen few of the clean files to be sure that this is indeed clean speech.
Hmm Is there any other way to confirm if dataset is correct?

>Yes,but not as bad..
Do you remember which data is bad?
If you are ok, let me know the detail of the issue at that time.

>Can you try with kernel_size 16, this is way more stable than the kernel_size 2.
OK Let me try.
I will share the result here but unfortunately I can not start immediately so please give me time.

>Also, did you inspect the log? How is the training loss?
This the last line of logs.
`Epoch 200: 100%|██████████| 1508/1508 [27:01<00:00,  1.08s/it, loss=5.354, v_num=0, val_loss=5.39]
`

**Comment by mpariente** [comment]
Which version of torch do you use? 
About the logs, check to whole training (with tensorboard) because DPRNN can diverge at some point..

**Comment by ntyoshi** [comment]
>Which version of torch do you use?

torch                   1.5.1
pytorch-lightning       0.8.5

>About the logs, check to whole training (with tensorboard) because DPRNN can diverge at some point..

I see. I'll share you the result later.

File: egs/wham/DPRNN/train.py [hunk 4]
```diff
@@ -66,31 +66,25 @@ def main(conf):
...
+                                 mode='min', save_top_k=5)
     early_stopping = False
     if conf['training']['early_stop']:
         early_stopping = EarlyStopping(monitor='val_loss', patience=30,
                                        verbose=1)
 
     # Don't ask GPU if they are not available.
     gpus = -1 if torch.cuda.is_available() else None
-    trainer = pl.Trainer(max_nb_epochs=conf['training']['epochs'],
+    trainer = pl.Trainer(max_epochs=conf['training']['epochs’],
```
Comment (jonashaag) [inline egs/wham/DPRNN/train.py:78]: The quote is broken (`’` instead of `'`)

**Comment by jonashaag** [comment]
When we merge this, shouldn't we also adapt all of the other copies of `train.py`?

File: egs/wham/DPRNN/train.py (hunk 4 above)
Comment (mpariente) [reply egs/wham/DPRNN/train.py:78]:
```suggestion
     trainer = pl.Trainer(max_epochs=conf['training']['epochs'], 
```

**Comment by mpariente** [comment]
Yes, this is the plan. But it seems @ntyoshi has issues with this script for now. 
I think we'll get rid of the `best_k_models.json` entirely, to be discussed.

**Comment by ntyoshi** [comment]
@mpariente 

>>Which version of torch do you use?

>torch 1.5.1
>pytorch-lightning 0.8.5

Which version did you use when you got the result [DPRNN's README.md](https://github.com/mpariente/asteroid/blob/master/egs/wham/DPRNN/README.md) says.

> >About the logs, check to whole training (with tensorboard) because DPRNN can diverge at some point..

>I see. I'll share you the result later.

Here is training loss and validation loss. It doesn't seem to be overfitting and underfitting but how was your result? Can you share me it?

image

Let me compare dataset on my environment and that on yours.
In my environment, the number of audio files is 336,000 and the total size of all is about 83 GB.
How about yours?

**Comment by mpariente** [comment]
Thanks for the answer. 
Did you launch the ks=16 version? Just to know if this was a terribly bad run or you have a problem with your data. Did you successfully train any model on WHAM?

**Comment by ntyoshi** [comment]
>Did you launch the ks=16 version? 

It's running on training. Please wait for a while.

>Did you successfully train any model on WHAM?

Not yet.

Can you tell me which torch version you used when you got the result DPRNN's README.md says?
And did WHAM dataset you used have 336,000 audio files / about 83GB?
I should use completely same environment.

**Comment by mpariente** [comment]
I used torch 1.5 I think (not so sure now)
About WHAM, which subset are you talking about? All of them? The wav8k? min
max? Please specify

**Comment by ntyoshi** [comment]
>About WHAM, which subset are you talking about? All of them? The wav8k? min
max? Please specify

wav8k/min and wav16k/min.
Let me know if you need more info. Thanks.

**Comment by mpariente** [comment]
In wav8k/min:
```bash
(pytorch3.6) [zaf67@vm-pul51 min]$ find . -iname *.wav | wc -l
16800
(pytorch3.6) [zaf67@vm-pul51 min]$ du -sh .
28G	.
```
That's about what you say. The 16K will weight double, which makes 336000 files for 83G.

**Comment by ntyoshi** [comment]
That's same as mine.

What do you think about loss I attached?
Is it similar as the result you got?

**Comment by mpariente** [comment]
No, it's very bad.. 
It's not at all what's expected

**Comment by mpariente** [comment]
Try to run ks=16, or convtasnet

**Comment by ntyoshi** [comment]
Quick question.
In [conf.yml](https://github.com/mpariente/asteroid/blob/master/egs/wham/DPRNN/local/conf.yml), there are the line below:

```
# Data config
data:
  train_dir: data/wav8k/min/tr/
  valid_dir: data/wav8k/min/cv/
```

Is wav16k used automatically in training process? 
If not, is there another file to specify wav16k?

**Comment by mpariente** [comment]
wav16k is not used by default. You can set the sampling rate through run.sh

**Comment by ntyoshi** [comment]
Oh ok. Let me know if there are other things to be changed from default.
The model that has ks=16 is still on process (I guess it is going to be finished 3 days later.)

**Comment by ntyoshi** [comment]
Here is the result of ks=16:

```
{
"si_sdr": 8.021911555643232,
"si_sdr_imp": 8.024758380181039,
"sdr": 9.0290907407023,
"sdr_imp": 8.679306945669248,
"sir": 16.63624848915427,
"sir_imp": 16.286464694121204,
"sar": 11.183047850847268,
"sar_imp": -137.99997502192846,
"stoi": 0.8159398188885097,
"stoi_imp": 0.13339052323656087
}
```

image

It is much better but is it not enough?

**Comment by jonashaag** [comment]
Why does the graph say validation SI-SDR is 17dB, but the Python dict you've pasted says it's -8dB?

**Comment by ntyoshi** [comment]
I pasted final_metrics.json. We need to check code.

**Comment by ntyoshi** [comment]
If you need more info about the result, please let me know.

**Comment by mpariente** [comment]
So, probably the best model which was retrieved is not the best model? 
Can you try to listen to some separated examples please?

**Comment by ntyoshi** [comment]
I took a look the examples, ex_1619 folder has metrics.json below:
```
{
"input_si_sdr": -0.0042400360107421875,
"input_sdr": 0.060265562506902626,
"input_sir": 0.06026556250692705,
"input_sar": 149.0144534901072,
"input_stoi": 0.7333606072538599,
"si_sdr": 15.965234756469727,
"sdr": 16.280655852431142,
"sir": 26.676765192082538,
"sar": 16.707819587583504,
"stoi": 0.9659945228146697,
"mix_path": "/home/ntyoshi/dataset/wham/wav8k/min/tt/mix_clean/423o0304_1.419_420c020x_-1.419.wav"
}
```
s1_estimate.wav and s2_estimate.wav sound clear.

**Comment by mpariente** [comment]
The final result seem weird then.. IDK sorry

**Comment by ntyoshi** [comment]
Should final result have best values of metrics?

**Comment by ntyoshi** [comment]
@mpariente 
Does your result on [README.md ](https://github.com/mpariente/asteroid/blob/fccf9d0c69b743eff0a2e636551c192a2194b774/egs/wham/DPRNN/README.md) mean the best result of metrics.json in /examples or final_metrics.json?

**Comment by ntyoshi** [comment]
@jonashaag 
The reason why final_metrics.json has -8dB seems to be related to the evaluation process or test data.
I checked all_metrics.csv and plot SI-SDR.
Then I could see that SI-SDRs have extremely low values if the number of examples are larger than about 1600 (Please see the graph I attached).
The mean value of SI-SDRs that are from 1 to 1600 was 14.3. 
When I listened to mixture.wav files in /examples that is larger than 1600, they were clear (s1_estimate.wav and s2_estimate.wav sound bad).

image

I will appreciate your any opinion and suggestion for next step to solve this issue.

**Comment by ntyoshi** [comment]
Let me update my situation.
In all_metrics.csv, 'si_sdr' seemed to depend on the audio files name on column 'mix_path'.
(If the audio file name is '050a0501_1.7783_442o030z_-1.7783.wav', the serial code of source audios are '050a0501' and '442o030z'.)
If the audio contains '44*******' source, 'si_sdr' tends to be bad.
When I calculated mean values depending on the case if mix_path contains '44*******' or not, the result was below:
- no  '44*******' source: 16.8
- one '44*******' source: 6.83
- two '44*******' source: -0.84

And the scatter plot that depends on the case if mix_path contains '44*******' or not was below:
image

I tried to listen a few of them (ex. '050a0501_1.7783_442o030z_-1.7783.wav') but I guess it sounds not wired...

@mpariente 
Have you ever heard such situations?
And if you have any suggestions or comments, please let me know.

**Comment by mpariente** [comment]
> @mpariente
> Does your result on [README.md ](https://github.com/mpariente/asteroid/blob/fccf9d0c69b743eff0a2e636551c192a2194b774/egs/wham/DPRNN/README.md) mean the best result of metrics.json in /examples or final_metrics.json?

`final_metrics.json`

Whoa, this is really surprising behavior (and great detective work!). I never heard about anything like that.  


>  tried to listen a few of them (ex. '050a0501_1.7783_442o030z_-1.7783.wav') but I guess it sounds not wired...

What do you mean by wired? 
I don't really know what these wsj0 prefix mean, did you look into it? Do the clean source sound good?

**Comment by ntyoshi** [comment]
>What do you mean by wired?
I don't really know what these wsj0 prefix mean, did you look into it? Do the clean source sound good?

I don't know about the meaning of the prefix 44*****. Is there any idea to look into it?
As long as I listen to them, 44***** audios are same quality as others like 22******, 05*******.

I'm wondering if I should rerun the script to create dataset. What do you think about it?

**Comment by ntyoshi** [comment]
>>I don't really know what these wsj0 prefix mean, did you look into it? 

>I don't know about the meaning of the prefix 44*****. Is there any ideas how to look into it?

The below document says the prefix means utterances. I guess they mean just serial codes.
[LDC93S6A](https://catalog.ldc.upenn.edu/docs/LDC93S6A/csrnov92.txt)

**Comment by mpariente** [comment]
I do think you should try to regenerate the data. Maybe only the test set. 
I don't remember, did you try evaluation on a pretrained-model? Take ConvTasNet and run evaluation? If you cannot reproduce the results, your data is wrong

**Comment by ntyoshi** [comment]
I will regenerate them. Thanks for the comment.

This model [(mpariente/DPRNNTasNet(ks=16)_WHAM!_sepclean)](https://zenodo.org/record/3903795#.XzOuRy2z2wQ) should be same parameter, right?
Let me try this model at first.

**Comment by mpariente** [comment]
Hey @ntyoshi, do you have some news?

**Comment by ntyoshi** [comment]
Yes.
I tried to test pre-trained model (mpariente/DPRNNTasNet(ks=16)_WHAM!_sepclean) but the result was wired. 
So I regenerated dataset and retried the test process. Then the result was almost same as that you were sharing. This meant regenerated dataset should be valid.
Next I am trying to train the model that has same parameters as (mpariente/DPRNNTasNet(ks=16)_WHAM!_sepclean). 
Let me share the result here.

Does the model have the same parameters as paper?

**Comment by mpariente** [comment]
Great to hear ! 

> Does the model have the same parameters as paper?

Yes, it does.

(I think you mean weird, not wired :upside_down_face: )

**Comment by ntyoshi** [comment]
@mpariente 
I see. Thank you for your great helps.

**Comment by mpariente** [comment]
Finally, this was fixed on master before. 
I'm closing this
